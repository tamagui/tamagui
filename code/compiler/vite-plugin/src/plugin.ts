import type { TamaguiOptions } from '@tamagui/static'
import type { Plugin } from 'vite'
import { transformWithEsbuild } from 'vite'
import { tamaguiExtractPlugin } from './extract'
import { Static, loadTamaguiBuildConfig, tamaguiOptions } from './loadTamagui'
import { fileURLToPath } from 'url'

const safeFileURLToPath = (fileURL?: string) => {
  if (!fileURL) return undefined;
  return fileURLToPath(fileURL);
};
const resolve = (name: string) => safeFileURLToPath(import.meta.resolve?.(name));

export function tamaguiPlugin({
  optimize,
  disableResolveConfig,
  ...tamaguiOptionsIn
}: TamaguiOptions & { optimize?: boolean; disableResolveConfig?: boolean } = {}):
  | Plugin
  | Plugin[] {
  const shouldAddCompiler = !!optimize
  let watcher
  let loaded = false

  // TODO temporary fix
  const enableNativeEnv = !!globalThis.__vxrnEnableNativeEnv

  const extensions = [
    `.web.mjs`,
    `.web.js`,
    `.web.jsx`,
    `.web.ts`,
    `.web.tsx`,
    '.mjs',
    '.js',
    '.mts',
    '.ts',
    '.jsx',
    '.tsx',
    '.json',
  ]

  async function load() {
    if (loaded) return
    loaded = true

    await loadTamaguiBuildConfig(tamaguiOptionsIn)

    if (tamaguiOptions!.disableWatchTamaguiConfig) {
      return
    }

    if (!Static) {
      throw new Error(`Not loaded`)
    }

    watcher = Static.watchTamaguiConfig({
      components: ['tamagui'],
      config: './src/tamagui.config.ts',
      ...tamaguiOptions,
    }).catch((err) => {
      console.error(` [Tamagui] Error watching config: ${err}`)
    })
  }

  const compatPlugins = [
    {
      name: 'tamagui-base-web-only',
      enforce: 'pre',

      async buildEnd() {
        await watcher?.then((res) => {
          res?.dispose()
        })
      },

      async transform(code, id) {
        if (id.includes('expo-linear-gradient')) {
          // fix expo-linear-gradient
          return transformWithEsbuild(code, id, {
            loader: 'jsx',
            jsx: 'automatic', // ðŸ‘ˆ
          })
        }
      },

      async config(_, env) {
        await load()

        if (!tamaguiOptions) {
          throw new Error(`No options loaded`)
        }

        return {
          environments: {
            client: {
              define: {
                'process.env.TAMAGUI_IS_CLIENT': JSON.stringify(true),
                'process.env.TAMAGUI_ENVIRONMENT': '"client"',
              },
            },
          },

          define: {
            // reanimated support
            _frameTimestamp: undefined,
            _WORKLET: false,
            __DEV__: `${env.mode === 'development'}`,
            'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || env.mode),
            'process.env.ENABLE_RSC': JSON.stringify(process.env.ENABLE_RSC || ''),
            'process.env.ENABLE_STEPS': JSON.stringify(process.env.ENABLE_STEPS || ''),
            'process.env.IS_STATIC': JSON.stringify(false),
            ...(env.mode === 'production' && {
              'process.env.TAMAGUI_OPTIMIZE_THEMES': JSON.stringify(true),
            }),
          },
          resolve:
            disableResolveConfig || enableNativeEnv
              ? {}
              : {
                  extensions,
                  alias: {
                    ...(tamaguiOptions.platform !== 'native' && {
                      'react-native/Libraries/Renderer/shims/ReactFabric':
                        resolve('@tamagui/proxy-worm'),
                      'react-native/Libraries/Utilities/codegenNativeComponent':
                        resolve('@tamagui/proxy-worm'),
                      'react-native-svg': resolve('@tamagui/react-native-svg'),
                      ...(!tamaguiOptions?.useReactNativeWebLite && {
                        'react-native': resolve('react-native-web'),
                      }),
                    }),
                  },
                },
        }
      },
    },

    {
      name: 'tamagui-rnw-lite-web-only',

      // enforce: 'pre',

      // resolveId(source) {
      //   const envName = this['environment']?.name as any // vite 5 + 6 compat
      //   if (isVite6 && envName !== 'client' && envName !== 'ssr') {
      //     return
      //   }

      //   if (source === 'react-native-svg') {
      //     return '@tamagui/react-native-svg'
      //   }

      //   console.log('cmon', source)

      //   if (tamaguiOptions?.useReactNativeWebLite) {
      //     if (/^react-native$/.test(source)) {
      //       return 'react-native-web'
      //     }
      //   } else {
      //     if (/^react-native$/.test(source)) {
      //       return '@tamagui/react-native-web-lite'
      //     }
      //   }
      // },

      config() {
        if (tamaguiOptions?.useReactNativeWebLite) {
          const rnwl = resolve(
            tamaguiOptions?.useReactNativeWebLite === 'without-animated'
              ? '@tamagui/react-native-web-lite/without-animated'
              : '@tamagui/react-native-web-lite'
          )
          const rnwlSS = resolve(
            '@tamagui/react-native-web-lite/dist/exports/StyleSheet/compiler/createReactDOMStyle'
          )

          return {
            resolve: {
              alias: [
                // fix reanimated issue not finding this
                {
                  find: /react-native.*\/dist\/exports\/StyleSheet\/compiler\/createReactDOMStyle/,
                  replacement: rnwlSS,
                },
                {
                  find: /^react-native$/,
                  replacement: rnwl,
                },
                {
                  find: /^react-native\/(.*)$/,
                  replacement: rnwl,
                },
                {
                  find: /^react-native-web$/,
                  replacement: rnwl,
                },
                {
                  find: /^react-native-web\/(.*)$/,
                  replacement: rnwl,
                },
              ],
            },
          }
        }
      },
    },
  ] satisfies Plugin[]

  if (shouldAddCompiler) {
    return [...compatPlugins, tamaguiExtractPlugin(tamaguiOptionsIn)]
  }

  return [...compatPlugins]
}
