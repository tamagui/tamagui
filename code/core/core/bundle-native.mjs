#!/usr/bin/env node

import { bundleNative } from '@tamagui/native-bundle'

// Bundle for production (native.cjs)
await bundleNative({
  entry: 'src/index.tsx',
  outDir: 'dist',
  fileName: 'native.cjs',
})

// Bundle for tests (test.native.cjs)
// Externalizes react-native so it can be mocked at runtime by vitest-react-native
await bundleNative({
  entry: 'src/index.tsx',
  outDir: 'dist',
  fileName: 'test.native.cjs',
  isTest: true,
  define: {
    'process.env.NODE_ENV': JSON.stringify('test'),
  },
})
