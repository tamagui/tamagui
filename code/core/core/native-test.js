Object.assign(module.exports, require('./dist/test'))
