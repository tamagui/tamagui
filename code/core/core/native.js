Object.assign(module.exports, require('./dist/native.cjs'))
