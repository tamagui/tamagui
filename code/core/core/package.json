{
  "name": "@tamagui/core",
  "version": "1.125.8",
  "source": "src/index.tsx",
  "main": "dist/cjs",
  "module": "dist/esm",
  "tamagui": {
    "bundle.native": "./src/index.tsx",
    "bundle.native.test": "./src/index.tsx"
  },
  "types": "./types/index.d.ts",
  "license": "MIT",
  "author": {
    "name": "Nate Wienert"
  },
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "tamagui-build",
    "watch": "tamagui-build --watch",
    "lint": "biome check src",
    "lint:fix": "biome check --write src",
    "clean": "tamagui-build clean",
    "clean:build": "tamagui-build clean:build"
  },
  "files": [
    "types",
    "dist",
    "src",
    "reset.css",
    "native.js",
    "native.d.ts",
    "native-test.js",
    "native-test.d.ts"
  ],
  "dependencies": {
    "@tamagui/react-native-media-driver": "workspace:*",
    "@tamagui/react-native-use-pressable": "workspace:*",
    "@tamagui/react-native-use-responder-events": "workspace:*",
    "@tamagui/use-event": "workspace:*",
    "@tamagui/web": "workspace:*"
  },
  "devDependencies": {
    "@tamagui/build": "workspace:*",
    "@testing-library/react": "^16.1.0",
    "csstype": "^3.0.10",
    "typescript": "^5.7.2",
    "vitest": "^3.0.5"
  },
  "exports": {
    "./package.json": "./package.json",
    "./reset.css": "./reset.css",
    ".": {
      "react-native": {
        "import": "./dist/esm/index.native.js",
        "require": "./dist/cjs/index.native.js"
      },
      "types": "./types/index.d.ts",
      "import": "./dist/esm/index.mjs",
      "require": "./dist/cjs/index.cjs",
      "default": "./dist/cjs/index.native.js"
    },
    "./native": {
      "types": "./types/index.d.ts",
      "import": "./dist/native.mjs",
      "require": "./dist/native.js",
      "react-native": "./dist/native.js"
    },
    "./native-test": {
      "types": "./types/index.d.ts",
      "import": "./dist/test.native.js",
      "require": "./dist/test.native.js",
      "react-native": "./dist/test.native.js"
    },
    "./inject-styles": {
      "types": "./types/inject-styles.d.ts",
      "import": "./dist/esm/inject-styles.mjs",
      "require": "./dist/cjs/inject-styles.cjs",
      "react-native": "./dist/cjs/inject-styles.native.js"
    },
    "./reactNativeTypes": {
      "types": "./types/reactNativeTypes.d.ts",
      "import": "./dist/esm/reactNativeTypes.mjs",
      "require": "./dist/cjs/reactNativeTypes.cjs",
      "react-native": "./dist/cjs/reactNativeTypes.native.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/tamagui/tamagui.git",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "a49cc7ea6b93ba384e77a4880ae48ac4a5635c14"
}
