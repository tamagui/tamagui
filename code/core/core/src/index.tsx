import { createMedia } from '@tamagui/react-native-media-driver'
import { useResponderEvents } from '@tamagui/react-native-use-responder-events'
import type {
  StackNonStyleProps,
  StackStyleBase,
  TamaDefer,
  TamaguiComponent,
  TamaguiElement,
  TamaguiTextElement,
  TextNonStyleProps,
  TextProps,
  TextStylePropsBase,
} from '@tamagui/web'
import {
  Stack as WebStack,
  Text as WebText,
  View as WebView,
  composeEventHandlers,
  createTamagui as createTamaguiWeb,
  setupHooks,
} from '@tamagui/web'
import React from 'react'

import { createOptimizedView } from './createOptimizedView'
import { getBaseViews } from './getBaseViews'
import { getRect } from './helpers/getRect'
import { measureLayout, useElementLayout } from './hooks/useElementLayout'
import type { RNTextProps, RNViewProps } from './reactNativeTypes'
import { usePressability } from './vendor/Pressability'
import { addNativeValidStyles } from './addNativeValidStyles'

// helpful for usage outside of tamagui
export {
  getBoundingClientRectAsync,
  measureLayout,
  useElementLayout,
  type LayoutEvent,
} from './hooks/useElementLayout'

// add newer style props based on react native version
addNativeValidStyles()

// adds extra types to View/Stack/Text:

type RNExclusiveViewProps = Omit<RNViewProps, keyof StackNonStyleProps>
export interface RNTamaguiViewNonStyleProps
  extends StackNonStyleProps,
    RNExclusiveViewProps {}

type RNTamaguiView = TamaguiComponent<
  TamaDefer,
  TamaguiElement,
  RNTamaguiViewNonStyleProps,
  StackStyleBase,
  {}
>

type RNExclusiveTextProps = Omit<RNTextProps, keyof TextProps>
export interface RNTamaguiTextNonStyleProps
  extends TextNonStyleProps,
    RNExclusiveTextProps {}

type RNTamaguiText = TamaguiComponent<
  TamaDefer,
  TamaguiTextElement,
  RNTamaguiTextNonStyleProps,
  TextStylePropsBase,
  {}
>

// re-exports all of @tamagui/web just adds hooks
export * from '@tamagui/web'
// fixes issues with TS saying internal type usage is breaking
// see https://discord.com/channels/909986013848412191/1146150253490348112/1146150253490348112
export * from './reactNativeTypes'

// automate using the react native media driver
export const createTamagui: typeof createTamaguiWeb = (conf) => {
  if (process.env.TAMAGUI_TARGET === 'native') {
    if (conf.media) {
      conf.media = createMedia(conf.media)
    }
  }
  return createTamaguiWeb(conf)
}

const baseViews = getBaseViews()

// setup internal hooks:

setupHooks({
  getBaseViews,

  setElementProps: (node) => {
    // web only
    if (node && !node['measure']) {
      // @ts-ignore
      node.measure ||= (callback) => measureLayout(node, null, callback)
      // @ts-ignore
      node.measureLayout ||= (relativeToNode, success) =>
        measureLayout(node as HTMLElement, relativeToNode, success)
      // @ts-ignore
      node.measureInWindow ||= (callback) => {
        setTimeout(() => {
          const { height, left, top, width } = getRect(node as HTMLElement)!
          callback(left, top, width, height)
        }, 0)
      }
    }
  },

  usePropsTransform(elementType, propsIn, stateRef, willHydrate) {
    if (process.env.TAMAGUI_TARGET === 'web') {
      const isDOM = typeof elementType === 'string'

      // replicate react-native-web functionality
      const {
        // remove event props handles by useResponderEvents
        onMoveShouldSetResponder,
        onMoveShouldSetResponderCapture,
        onResponderEnd,
        onResponderGrant,
        onResponderMove,
        onResponderReject,
        onResponderRelease,
        onResponderStart,
        onResponderTerminate,
        onResponderTerminationRequest,
        onScrollShouldSetResponder,
        onScrollShouldSetResponderCapture,
        onSelectionChangeShouldSetResponder,
        onSelectionChangeShouldSetResponderCapture,
        onStartShouldSetResponder,
        onStartShouldSetResponderCapture,

        // android
        collapsable,
        focusable,

        // deprecated,
        accessible,
        accessibilityDisabled,

        onLayout,
        hrefAttrs,

        ...plainDOMProps
      } = propsIn

      if (willHydrate || isDOM) {
        useElementLayout(stateRef, !isDOM ? undefined : (onLayout as any))
        useResponderEvents(stateRef, !isDOM ? undefined : propsIn)
      }

      if (isDOM) {
        // TODO move into getSplitStyles
        if (plainDOMProps.href && hrefAttrs) {
          const { download, rel, target } = hrefAttrs
          if (download != null) {
            plainDOMProps.download = download
          }
          if (rel) {
            plainDOMProps.rel = rel
          }
          if (typeof target === 'string') {
            plainDOMProps.target = target.charAt(0) !== '_' ? `_${target}` : target
          }
        }
        return plainDOMProps
      }
    }
  },

  useEvents(viewProps, events, { pseudos }, setStateShallow, staticConfig) {
    if (process.env.TAMAGUI_TARGET === 'native') {
      if (events) {
        if (events.onFocus) {
          viewProps['onFocus'] = events.onFocus
        }
        if (events.onBlur) {
          viewProps['onBlur'] = events.onBlur
        }
      }

      if (staticConfig.isInput) {
        if (events) {
          const { onPressIn, onPressOut, onPress } = events
          const inputEvents = {
            onPressIn,
            onPressOut: onPressOut || onPress,
          }
          if (onPressOut && onPress) {
            // only supports onPressIn and onPressOut so combine them
            inputEvents.onPressOut = composeEventHandlers(onPress, onPressOut)
          }
          Object.assign(viewProps, inputEvents)
        }
      } else {
        // use Pressability to get smooth unPress when you press + hold + move out
        // only ever create once, use .configure() to update later
        if (events && viewProps.hitSlop) {
          events.hitSlop = viewProps.hitSlop
        }

        // note we do events checks more than we should because we need this hook to always run
        const pressability = usePressability(events)

        if (events) {
          if (process.env.NODE_ENV === 'development') {
            if (viewProps['debug']) {
              console.info(
                `Checking for press ${!!events.onPress} then applying pressability props: ${Object.keys(
                  pressability || {}
                )}`
              )
            }
          }

          if (events.onPress) {
            for (const key in pressability) {
              const og = viewProps[key]
              const val = pressability[key]
              viewProps[key] =
                og && !dontComposePressabilityKeys[key]
                  ? composeEventHandlers(og, val)
                  : val
            }
          }
        }
      }
    }
  },

  // attempt at properly fixing RN input, but <Pressable><TextInput /> just doesnt work on RN
  ...(process.env.TAMAGUI_TARGET === 'native' && {
    useChildren(elementType, children, viewProps) {
      if (process.env.NODE_ENV === 'test') {
        // test mode - just use regular views since optimizations cause weirdness
        return
      }

      if (elementType === baseViews.View) {
        // optimize view
        return createOptimizedView(children, viewProps, baseViews)
      }

      if (process.env.TAMAGUI_OPTIMIZE_NATIVE_VIEWS) {
        if (elementType === baseViews.Text) {
          // further optimize by not even caling elementType.render
          viewProps.children = children
          return React.createElement('RCTText', viewProps)
        }
      }
    },
  }),
})

const dontComposePressabilityKeys = {
  onClick: true,
}

// overwrite web versions:
// putting at the end ensures it overwrites in dist/cjs/index.js
export const View = WebView as any as RNTamaguiView
export const Stack = WebStack as any as RNTamaguiView
export const Text = WebText as any as RNTamaguiText

// easily test type declaration output and if it gets messy:

// export const X = styled(WebView, {
//   variants: {
//     abc: {
//       true: {},
//     },
//   } as const,
// })

// export const Y = styled(X, {
//   variants: {
//     zys: {
//       true: {},
//     },
//   } as const,
// })

// export const Z = styled(Y, {
//   variants: {
//     xxx: {
//       true: {},
//     },
//   } as const,
// })

// export const A = styled(Z, {
//   variants: {} as const,
// })

// const zz = <A />

// const variants = {
//   fullscreen: {
//     true: {},
//   },
//   elevation: {
//     '...size': () => ({}),
//     ':number': () => ({}),
//   },
// } as const

// export const YStack = styled(View, {
//   flexDirection: 'column',
//   variants,
// })

// import { TextInput } from 'react-native'
// export const InputFrame = styled(
//   TextInput,
//   {
//     name: 'Input',
//     backgroundColor: 'green',

//     variants: {
//       // unstyled: {
//       //   false: {},
//       // },

//       size: {
//         '...size': () => ({}),
//       },

//       // disabled: {
//       //   ':boolean': () => ({})
//       // },
//     } as const,

//     // defaultVariants: {
//     //   unstyled: process.env.TAMAGUI_HEADLESS === '1' ? true : false,
//     // },
//   },
//   {
//     isText: true,
//     accept: {
//       placeholderTextColor: 'color',
//     },
//   }
// )

// export const StyledInputFrame = styled(InputFrame, {
//   fontSize: 16,
//   fontFamily: '$silkscreen',
//   color: '$color5',
//   minWidth: 0,
//   borderWidth: 0,
//   borderColor: 'transparent',

//   variants: {
//     unset: {
//       false: {
//         borderWidth: 2,
//         py: 12,
//         px: 14,
//         borderRadius: 6,
//         bg: '$color3',
//         focusStyle: {
//           bg: '$color4',
//           margin: 0,
//         },
//       },
//     },
//   } as const,

//   defaultVariants: {
//     unset: false,
//   },
// })

// export const StyledStyledInputFrame = styled(
//   StyledInputFrame,
//   {
//     fontSize: 16,
//     fontFamily: '$silkscreen',
//     color: '$color5',
//     minWidth: 0,
//     borderWidth: 0,
//     borderColor: 'transparent',

//     variants: {
//       unset: {
//         false: {
//           borderWidth: 2,
//           py: 12,
//           px: 14,
//           borderRadius: 6,
//           bg: '$color3',
//           focusStyle: {
//             bg: '$color4',
//             margin: 0,
//           },
//         },
//       },
//     } as const,

//     defaultVariants: {
//       unset: false,
//     },
//   },
//   {
//     inlineProps: new Set(['id', 'testID']),
//   }
// )
// export const DepthTest = () => <StyledStyledInputFrame placeholder="" />
