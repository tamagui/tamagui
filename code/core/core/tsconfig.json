{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "composite": true,
    "strict": true
  },
  "exclude": [
    "./native.d.ts",
    "./native-test.d.ts",
    "./types",
    "types",
    "dist",
    "**/__tests__",
    "vite.bundle.config.ts"
  ],
  "references": [
    {
      "path": "../react-native-use-pressable"
    },
    {
      "path": "../react-native-use-responder-events"
    },
    {
      "path": "../use-event"
    },
    {
      "path": "../web"
    }
  ]
}
