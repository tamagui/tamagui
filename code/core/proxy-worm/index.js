module.exports = worm(true)

function worm(root = false) {
  const obj = root
    ? {
        get default() {
          return worm()
        },
      }
    : (() => {
        //  dev warning in debug mode
        if (
          process.env.NODE_ENV === 'development' &&
          process.env.DEBUG?.startsWith('tamagui')
        ) {
          console.warn(`

This has been excluded via Tamagui!
Check "excludeReactNativeWebExports" setting and include it to fix.

${new Error().stack}

`)
        }

        return worm()
      })

  obj.displayName = `ProxyWorm - Check excludeReactNativeWebExports`
  obj._isProxyWorm = true

  // reanimated tries to find component like things
  obj.prototype = obj.prototype || {}
  obj.prototype.isReactComponent = true

  return new Proxy(obj, {
    get(_, key) {
      if (Reflect.has(obj, key)) return Reflect.get(obj, key)
      return worm()
    },
    apply() {
      return worm()
    },
  })
}
