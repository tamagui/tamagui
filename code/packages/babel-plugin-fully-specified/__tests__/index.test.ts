import path from 'node:path'
import { describe, expect, test } from 'vitest'
import { type TransformOptions, transform, transformFileSync } from '@babel/core'

import plugin, { type FullySpecifiedOptions } from '../src'

/** A helper function to get the default transform options for the plugin to test. */
const getTransformOptions = ({
  pluginOptions,
}: { pluginOptions?: Partial<FullySpecifiedOptions> } = {}) => ({
  plugins: [[plugin, { ensureFileExists: false, ...pluginOptions }]],
})

/** A test helper for calling Babel transform with the plugin to test and some default options. */
function getTransformResult(
  input,
  {
    transformOptions,
    pluginOptions,
  }: {
    transformOptions?: TransformOptions
    pluginOptions?: Partial<FullySpecifiedOptions>
  } = {}
) {
  return transform(input, {
    filename: 'myFile.js',
    configFile: false,
    ...transformOptions,
    ...getTransformOptions({ pluginOptions }),
  })
}

describe('local imports', () => {
  test('named import', () => {
    const example1 = "import { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import { foo } from "./foo.mjs";')

    const example2 = "import { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import { foo as bar } from "./foo.mjs";'
    )
  })

  test('default import', () => {
    const example1 = "import defaultFoo from './foo'"
    expect(getTransformResult(example1)?.code).toBe('import defaultFoo from "./foo.mjs";')

    const example2 = "import defaultFoo, { foo } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'import defaultFoo, { foo } from "./foo.mjs";'
    )
  })

  test('namespace import', () => {
    const example = "import * as fooModule from './foo'"
    expect(getTransformResult(example)?.code).toBe(
      'import * as fooModule from "./foo.mjs";'
    )
  })

  test('side effects only import', () => {
    const example = "import './foo'"
    expect(getTransformResult(example)?.code).toBe('import "./foo.mjs";')
  })

  test('dynamic import()', () => {
    const example = "const foo = await import('./foo')"
    expect(getTransformResult(example)?.code).toBe(
      'const foo = await import("./foo.mjs");'
    )
  })
})

describe('local re-exports', () => {
  test('named', () => {
    const example1 = "export { foo } from './foo'"
    expect(getTransformResult(example1)?.code).toBe('export { foo } from "./foo.mjs";')

    const example2 = "export { foo as bar } from './foo'"
    expect(getTransformResult(example2)?.code).toBe(
      'export { foo as bar } from "./foo.mjs";'
    )
  })

  test('default', () => {
    const example1 = "export { default as fooDefault } from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export { default as fooDefault } from "./foo.mjs";'
    )
  })

  test('namespace', () => {
    const example1 = "export * as fooModule from './foo'"
    expect(getTransformResult(example1)?.code).toBe(
      'export * as fooModule from "./foo.mjs";'
    )

    const example2 = "export * from './foo'"
    expect(getTransformResult(example2)?.code).toBe('export * from "./foo.mjs";')
  })
})

describe('transforming actual files', () => {
  test('multiple extensions exists', () => {
    const { code } =
      transformFileSync(
        path.join(__dirname, 'fixtures', 'multiple-extensions-exists', 'test.mjs'),
        getTransformOptions({
          pluginOptions: { ensureFileExists: true },
        })
      ) || {}

    expect(code).toMatchSnapshot()
  })
})
