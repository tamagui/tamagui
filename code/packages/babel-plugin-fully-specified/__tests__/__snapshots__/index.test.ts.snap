// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforming actual files > multiple extensions exists 1`] = `
"// Expected to be transformed to use the \`.mjs\` extension, which is same as this file.
import { foo, bar } from "./modules/module.mjs";"
`;
